### PR TITLE
[busboy] Update to version 1.3

### PR DESCRIPTION
**Pull request by martin-badin**
Discussion: #57917 
Documentation: https://github.com/mscdex/busboy

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test busboy`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <https://github.com/mscdex/busboy>
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

File: types/async-busboy/async-busboy-tests.ts [hunk 1]
```diff
@@ -8,20 +8,18 @@ type Result = asyncBusboy.Result;
 type OnFileResult = asyncBusboy.OnFileResult;
 type FileReadStream = asyncBusboy.FileReadStream;
 
-const req = null as any as IncomingMessage;
+const req = (null as any) as IncomingMessage;
```
Comment (peterblazejewicz) [inline types/async-busboy/async-busboy-tests.ts:11]:
```suggestion
declare const req: IncomingMessage;
```

File: types/busboy/index.d.ts [hunk 2]
```diff
@@ -1,114 +1,145 @@
...
+
+        /**
+         * For multipart forms, the max number of header key-value pairs to parse.
+         *
+         * @default 2000 (same as node's http module)
+         */
+        headerPairs?: number | undefined;
+    }
+
+    interface Config {
```
Comment (peterblazejewicz) [inline types/busboy/index.d.ts:65]: can you left alias for backward compatibility (like `type BusboyConfig = Config;`)
the same for all other update/rewritten implementation?

Comment (martin-badin) [reply types/busboy/index.d.ts:65]: yes, sure

File: types/busboy/v0/busboy-tests.ts [hunk 3]
```diff
@@ -0,0 +1,261 @@
+import Busboy = require('busboy');
```
Comment (peterblazejewicz) [inline types/busboy/v0/busboy-tests.ts:1]:
```suggestion
import busboy = require('busboy');
```
this is not how they name things:
https://github.com/mscdex/busboy/compare/v0.3.1...v1.0.0#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R101
so:
```js
const busboy = require('busboy');
...
const bb = busboy({...});

File: types/busboy/v0/busboy-tests.ts [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+// test type exports
+type Constructor = Busboy.BusboyConstructor;
+type Config = Busboy.BusboyConfig;
+type Headers = Busboy.BusboyHeaders;
+type BB = Busboy.Busboy;
+type Events = Busboy.BusboyEvents;
+
+new Busboy({}); // $ExpectError
+const busboy = Busboy({ headers: { 'content-type': 'foo' } }); // $ExpectType Busboy
+new Busboy({ headers: { 'content-type': 'foo' } }); // $ExpectType Busboy
```
Comment (peterblazejewicz) [inline types/busboy/v0/busboy-tests.ts:12]: is this even possible under v1? It looks only factory method is exported, returning instance or throwing an error.

File: types/busboy/v0/busboy-tests.ts (hunk 3 above)
Comment (peterblazejewicz) [reply types/busboy/v0/busboy-tests.ts:1]: Oh, sorry, I was mislead by versioning.
Please do not leave v0, it's 3 years old (in source). This package depends on Node, so it impacts Node CI tests on DT (each version adds a time for Node CI tests)

File: types/busboy/v0/busboy-tests.ts (hunk 4 above)
Comment (peterblazejewicz) [reply types/busboy/v0/busboy-tests.ts:12]: forget this one, I've missed versioning

File: types/multy/index.d.ts [hunk 5]
```diff
@@ -4,17 +4,17 @@
...
-declare module "koa" {
+declare module 'koa' {
     interface Request {
         body: any;
     }
 }
 
 declare namespace multy {
-    type Options = busboy.BusboyConfig;
+    type Options = busboy.Config;
```
Comment (peterblazejewicz) [inline types/multy/index.d.ts:17]: my comment in `v0` about aliases was intended for this use-case. 
Provide alias, if that is enough for backward compatible DT types.
if you remove v0, this package will require `package.json`, as it is 5 years old and
orks only with v0.2 of

File: types/koa-joi-router/index.d.ts [hunk 6]
```diff
@@ -60,9 +60,7 @@ declare namespace createRouter {
         meta?: any;
     }
 
-    interface MultipartOptions extends Omit<busboy.BusboyConfig, 'headers'> {
-        headers?: http.IncomingHttpHeaders;
-    }
+    type MultipartOptions = busboy.Config;
```
Comment (peterblazejewicz) [inline types/koa-joi-router/index.d.ts:63]: _ditto_ about aliases

File: types/express-fileupload/index.d.ts [hunk 7]
```diff
@@ -8,7 +8,7 @@
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 
 import type { RequestHandler } from 'express';
-import type { BusboyConfig } from 'busboy';
+import type { Config as BusboyConfig } from 'busboy';
```
Comment (peterblazejewicz) [inline types/express-fileupload/index.d.ts:11]: ditto about aliases

File: types/connect-busboy/index.d.ts [hunk 8]
```diff
@@ -14,8 +14,7 @@ import * as http from 'http';
 declare function connectBusboy(options?: connectBusboy.ConnectBusboyOptions): RequestHandler;
 
 declare namespace connectBusboy {
-    interface ConnectBusboyOptions extends Omit<busboy.BusboyConfig, 'headers'> {
-        headers?: http.IncomingHttpHeaders;
+    interface ConnectBusboyOptions extends busboy.Config {
```
Comment (peterblazejewicz) [inline types/connect-busboy/index.d.ts:17]: _ditto_ about  aliases, also use direct type import:
```
import { BusboyConfig }  from 'busboy';
...
interface ConnectBusboyOptions extends BusboyConfig {
```

File: types/connect-busboy/index.d.ts [hunk 9]
```diff
@@ -1,4 +1,4 @@
-// Type definitions for connect-busboy 0.0
+// Type definitions for connect-busboy 1.0
```
Comment (peterblazejewicz) [inline types/connect-busboy/index.d.ts:1]: can we drop TS 2.3 requirement header?

File: types/async-busboy/async-busboy-tests.ts [hunk 10]
```diff
@@ -8,20 +8,18 @@ type Result = asyncBusboy.Result;
...
 asyncBusboy(req, { highWaterMark: 10 }); // $ExpectType Promise<Result>
 
 const onFileResultPromise = asyncBusboy(req, {
-    onFile: (fieldname, file, filename, encoding, mimetype) => {
-        fieldname; // $ExpectType string
-        file; // $ExpectType Readable
-        filename; // $ExpectType string
-        encoding; // $ExpectType string
-        mimetype; // $ExpectType string
+    onFile: (name, stream, info) => {
```
Comment (peterblazejewicz) [inline types/async-busboy/async-busboy-tests.ts:19]: is that correct?
https://github.com/m4nuC/async-busboy/blob/master/index.js#L111
https://github.com/m4nuC/async-busboy#asyncawait-using-custom-onfile-handler-ie-no-temp-files

**Comment by peterblazejewicz** [comment]
@martin-badin you can either remove 'v0' completely (but there could be impact on 3rd party types here on DT or add '@types/busboy' to allowed types white list, so you can cleanly handle 5 years old changes in extreme use case here

File: types/async-busboy/async-busboy-tests.ts (hunk 10 above)
Comment (martin-badin) [reply types/async-busboy/async-busboy-tests.ts:19]: this is one of the questions I want to ask you. How I can use an older version of the package (busboy) in this package (async-busboy)? The package still uses older version https://github.com/m4nuC/async-busboy/blob/master/package.json#L23

File: types/busboy/v0/busboy-tests.ts (hunk 3 above)
Comment (martin-badin) [reply types/busboy/v0/busboy-tests.ts]: I removed v0

File: types/connect-busboy/index.d.ts (hunk 9 above)
Comment (martin-badin) [reply types/connect-busboy/index.d.ts:1]: I think that yes

File: types/multy/package.json [hunk 11]
```diff
@@ -0,0 +1,6 @@
+{
+    "private": true,
+    "dependencies": {
+        "@types/busboy": "^0.3.1"
```
Comment (peterblazejewicz) [inline types/multy/package.json:4]:
```suggestion
        "@types/busboy": "^0.2"
```
the package is 5 years old, let's try first proper versioning:
https://github.com/EduardoRFS/multy/blob/master/package.json#L28

**Comment by peterblazejewicz** [comment]
cool, ping me when upstream change gets approved

**Comment by martin-badin** [comment]
@peterblazejewicz PR was merged https://github.com/microsoft/DefinitelyTyped-tools/pull/373 but the build still failed. I don't know what is wrong.

**Comment by peterblazejewicz** [comment]
> @peterblazejewicz PR was merged [microsoft/DefinitelyTyped-tools#373](https://github.com/microsoft/DefinitelyTyped-tools/pull/373) but the build still failed. I don't know what is wrong.

this looks like error in 3rd party library tests, the comment shows that similar happened under 4.1:
> // can't assert via ExpectType because TS4.1 detects OnFileResult as Pick<Result, "fields">

You should be able to figure out what's wrong running for this library only from your PR branch:
```bash
npm test async-busboy
```

**Comment by martin-badin** [comment]
>

**Comment by martin-badin** [comment]
@peterblazejewicz The problem is with `@types/busboy`. `async-busboy` is tested with the newest version of the library, not with the version defined in the package.json

**Comment by peterblazejewicz** [comment]
oh, I see. I think this is my fault (giving wrong suggestion about versioning), everything else is ok, give me 5 m

**Comment by peterblazejewicz** [comment]
so that's my fault, mixed use-case with old types, the v0 approach is the good one here, the patch I tested locally:
https://gist.github.com/peterblazejewicz/a485fcddb9b732fc356abe25af6ee290

**Comment by martin-badin** [comment]
@peterblazejewicz Thank you for the patch.

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@martin-badin thx!

**Comment by martin-badin** [comment]
Ready to merge

**Pull request merged**